source/modulo_components/test/cpp/test_component_interface.cpp

```diff
@@ -0,0 +1,65 @@
+#include <gtest/gtest.h>
+#include <rclcpp/node.hpp>
+#include <rclcpp/publisher.hpp>
+#include <rclcpp/node_options.hpp>
+
+#include "modulo_components/ComponentInterface.h"
+#include "modulo_new_core/EncodedState.hpp"
+
+namespace modulo_components {
+class TestComponentInterface : public ::testing::Test {
+protected:
+  static void SetUpTestSuite() {
+    rclcpp::init(0, nullptr);
+    if (component_ == nullptr) {
+      component_ =
+          new ComponentInterface<rclcpp::Node, rclcpp::Publisher<modulo_new_core::EncodedState>>(rclcpp::NodeOptions());
+    }
+  }
+
+  void TearDown() override {
+    component_->predicates_.clear();
+  }
+
+  static ComponentInterface<rclcpp::Node, rclcpp::Publisher<modulo_new_core::EncodedState>>* component_;
+};
+
+ComponentInterface<rclcpp::Node, rclcpp::Publisher<modulo_new_core::EncodedState>>
+    * TestComponentInterface::component_ = nullptr;
```

Comment:
Why do you use a static instance here and not just a regular property? It's a bit weird to read like this, and I would have thought this wouldn't be necessary for the friend behaviour.

Also, you shouldn't use raw pointer. Since we are internally using `std::shared_from_this` in the class, it's imperative that you actually hold a reference to the instance through a smart pointer, otherwise things can get weird (with no static or run-time assertion possible).

Finally, the intention was to pass a publisher typename symbol (e.g. Publisher or lifecycle::Publisher), rather than a class with a message type like `rclcpp::Publisher<modulo_new_core::EncodedState>`. So, overall, I feel like it should be:

```suggestion
  std::shared_ptr<ComponentInterface<rclcpp::Node, rclcpp::Publisher>> component_ = nullptr;
};
```

Let me know if I'm missing something

Reply:
Alright so the problem is that `rclcpp::Publisher` are actually not instantiable as they are template. There is a `rclcpp::GenericPublisher` that is not templated. I was actually using that one before switching to the one from Dominic as I thought it was raising an error and forgot to change back.

Reply:
Regarding the static pointer I just followed the recommendation on the link you shared:
https://github.com/google/googletest/blob/main/docs/advanced.md#sharing-resources-between-tests-in-the-same-test-suite

I believe this will be the right approach when we use the node to test publishers and such. However, I changed for `std::unique_ptr` which works fine.